addons/html_editor/static/src/main/hint_plugin.js

```diff
@@ -64,26 +43,35 @@ export class HintPlugin extends Plugin {
     updateHints() {
         const selectionData = this.dependencies.selection.getSelectionData();
         const editableSelection = selectionData.editableSelection;
-        if (this.hint) {
-            const blockEl = closestBlock(editableSelection.anchorNode);
-            this.removeHint(this.hint);
-            this.removeHint(blockEl);
-        }
+        const blockEl = closestBlock(editableSelection.anchorNode);
+        this.removeHint(blockEl);
+        this.clearHints();
         if (editableSelection.isCollapsed) {
-            for (const hint of this.getResource("hints")) {
-                if (hint.selector) {
-                    const el = closestBlock(editableSelection.anchorNode);
-                    if (el && el.matches(hint.selector) && !isProtected(el) && isEmptyBlock(el)) {
-                        this.makeHint(el, hint.text);
-                        this.hint = el;
+            const hints = this.getResource("hints");
+            for (const cb of this.getResource("hint_targets_providers")) {
+                for (const target of cb(selectionData, this.editable)) {
+                    const nodeHint = hints.find((h) => target.matches(h.selector))?.text;
+                    if (
+                        target &&
+                        !target.classList.contains("o-we-hint") &&
+                        nodeHint &&
+                        isEmptyBlock(target)
+                    ) {
+                        this.makeHint(target, nodeHint);
                     }
-                } else {
-                    const target = hint.target(selectionData, this.editable);
-                    // Do not replace an existing empty block hint by a temp hint.
-                    if (target && !target.classList.contains("o-we-hint")) {
-                        this.makeHint(target, hint.text);
-                        this.hint = target;
-                        return;
+                }
+            }
+            if (!selectionData.documentSelectionIsInEditable) {
+                return;
+            }
+            for (const hint of hints) {
+                if (hint.selector) {
+                    if (
+                        blockEl?.matches(hint.selector) &&
+                        !isProtected(blockEl) &&
+                        isEmptyBlock(blockEl)
+                    ) {
+                        this.makeHint(blockEl, hint.text);
                     }
                 }
```

Comment:
This is not really needed anymore as it can make use of `hint_targets_providers`. If you define a target provider that returns `closestBlock(editableSelection.anchorNode)`, and add an `!isProtected(target)` in the `if` in the `for` lool above.